src/core/mapping/base_mapper.cc

```diff
@@ -941,34 +979,36 @@ void BaseMapper::legate_select_sources(const MapperContext ctx,
   std::vector<std::pair<PhysicalInstance, uint32_t /*bandwidth*/>> band_ranking;
   for (uint32_t idx = 0; idx < sources.size(); idx++) {
     const PhysicalInstance& instance = sources[idx];
-    Memory location                  = instance.get_location();
-    if (location.address_space() == local_node) {
-      if (!all_local) {
-        source_memories.clear();
-        band_ranking.clear();
-        all_local = true;
-      }
-    } else if (all_local)  // Skip any remote instances once we're local
-      continue;
-    auto finder = source_memories.find(location);
-    if (finder == source_memories.end()) {
-      affinity.clear();
-      machine.get_mem_mem_affinity(
-        affinity, location, destination_memory, false /*not just local affinities*/);
-      uint32_t memory_bandwidth = 0;
-      if (!affinity.empty()) {
-        assert(affinity.size() == 1);
-        memory_bandwidth = affinity[0].bandwidth;
-      }
-      source_memories[location] = memory_bandwidth;
-      band_ranking.push_back(std::pair<PhysicalInstance, uint32_t>(instance, memory_bandwidth));
-    } else
-      band_ranking.push_back(std::pair<PhysicalInstance, uint32_t>(instance, finder->second));
+    add_instance_to_band_ranking(instance,
+                                 local_node,
+                                 all_local,
+                                 source_memories,
+                                 band_ranking,
+                                 affinity,
+                                 destination_memory,
+                                 machine);
   }
-  // If there aren't any sources (for example if there are some collective views
-  // to choose from, not yet in this branch), just return nothing and let the
-  // runtime pick something for us.
-  if (band_ranking.empty()) { return; }
+
+  if (!all_local)  // no need to go through the collective instanes if there
```

Comment:
I feel like we'd want to go through the collective views, even if we found a local stand-alone instance `I`. There might be another instance in a collective view that's actually closer to the target than `I`.

Reply:
It also might be more expensive to figure out which one you want. If you find a local instance in the same memory as the target, you aren't going to do better than that digging through a collective instance.

Reply:
Note that "local" in this code refers to "same address space", not necessarily "directly on the target memory", so there is a (small) possibility that we are trying to source an instance on GPU 0, there is a standalone instance `I` on GPU 4 on the same node but a different NUMA domain, and a collective view contains an instance `J` that's on GPU 1, which is on the same NUMA domain as GPU 0 (or even directly on GPU 0!).

I don't like having an arbitrary fork in the logic in this case, where a whole class of possibly same-address-space instances are ignored, but only if they're part of a collective view. Unless we're trying hard to avoid a call to `find_instances_nearest_memory`, because it's expected to be prohibitively expensive (e.g. if it triggers remote queries) -- @lightsighter ?

A safe optimization that we could do instead in the `all_local` case is to only consider instances within the `CollectiveView` that are on the local address space. There doesn't seem to be a way to filter based on that characteristic on `CollectiveView`, so either we can add that API call, or find all the local memories and iterate over those using `find_instances_in_memory`. Ideally, however, `CollectiveView` would just provide an iterator over contained instances. This would be more practical in the general case, because then a mapper is free to prioritize however it wants, including on things other than memory channel latency/bandwidth, e.g. instance layout.

Reply:
This conversation was missing some context, see my updated comment below.